numba_cuda/numba/cuda/dispatcher.py

```diff
@@ -265,17 +266,28 @@ def maybe_link_nrt(self, link, tgt_ctx, asm):
         if not tgt_ctx.enable_nrt:
             return
 
-        all_nrt = "|".join(self.NRT_functions)
-        pattern = (
-            r"\.extern\s+\.func\s+(?:\s*\(.+\)\s*)?("
-            + all_nrt
-            + r")\s*\([^)]*\)\s*;"
-        )
-
-        nrt_in_asm = re.findall(pattern, asm)
-
-        basedir = os.path.dirname(os.path.abspath(__file__))
-        if nrt_in_asm:
+        nrt_in_asm = lambda asm: any(fn in asm for fn in self.NRT_functions)
+        link_nrt = nrt_in_asm(asm)
+        if not link_nrt:
+            for file in link:
+                if isinstance(file, LinkableCode):
+                    if file.nrt:
+                        link_nrt = True
+                        break
+                    elif isinstance(file, (CUSource, PTXSource)):
+                        asm = file.data.decode("utf-8")
```

Comment:
Note that user can still specify non-binary input which can lead to `asm` object undefined here.

```suggestion
                if isinstance(file, LinkableCode):
                    if file.nrt:
                        link_nrt = True
                        break
                    elif isinstance(file, (CUSource, PTXSource)):
                        asm = file.data.decode("utf-8")
                    else:
                        asm = ""
```